### PR TITLE
pid1,repart,growfs: make libcryptsetup a dlopen() style dep

### PR DESCRIPTION
**Pull request by poettering**
Let's make libcryptsetup a dlopen() style dep for PID 1 (i.e. for
    RootImage= and stuff), systemd-growfs and systemd-repart. (But leave to
    be a regulra dep in systemd-cryptsetup, systemd-veritysetup and
    systemd-homed since for them the libcryptsetup support is not auxiliary
    but pretty much at the core of what they do.)
    
    This should be useful for container images that want systemd in the
    payload but don't care for the cryptsetup logic since dm-crypt and stuff
    isn't available in containers anyway.
    
Fixes: #8249

### CODE REVIEW

**Comment by bluca** [comment]
I'm a bit concerned about perf impact, given as far as I can see pid1 will not dlopen, but the forked processes that start each service will have to do it themselves, so the cost has to be incurred every time as far as I can see, and this is in my critical path. If and only if future measurements show a meaningful impact, would it be acceptable if I sent a follow-up to add a meson option to reinstate linking as a disabled-by-default alternative?

**Comment by keszybz** [comment]
LGTM.

> I'm a bit concerned about perf impact, given as far as I can see pid1 will not dlopen, but the forked processes that start each service will have to do it themselves, so the cost has to be incurred every time as far as I can see, and this is in my critical path.

I'm also concerned about long-term maintainability. We copy the function headers and so-version suffix into our code. If things change, we won't notice. So longer-term I think we'll need to change how we do this. OTOH, short-term this shaves the deps in very important ways, so there's a clear immediate benefit for users.

> If and only if future measurements show a meaningful impact, would it be acceptable if I sent a follow-up to add a meson option to reinstate linking as a disabled-by-default alternative?

Can you first do the measurements? That'd be super interesting on its own. It is possible that this doesn't actually affect performance, and it'd be great to have a confirmation. But before you start working on a patch, I'd like to discuss how to we handle those issues in all the dlopen paths.

**Comment by keszybz** [comment]
centos7 failed in TEST-50-DISSECT. It could be potentially related.

<details>

```
19:49:11          Starting [0;1;39mCreate Volatile Files and Directories[0m...
19:49:11 [[0;32m  OK  [0m] Finished [0;1;39mCreate Volatile Files and Directories[0m.
19:49:11 [[0;32m  OK  [0m] Finished [0;1;39mdracut pre-udev hook[0m.
19:49:11          Starting [0;1;39mRule-based Manage…for Device Events and Files[0m...
19:49:11 [[0;32m  OK  [0m] Started [0;1;39mRule-based Manager for Device Events and Files[0m.
19:49:11          Starting [0;1;39mColdplug All udev Devices[0m...
19:49:11          Mounting [0;1;39mKernel Configuration File System[0m...
19:49:11 [[0;32m  OK  [0m] Mounted [0;1;39mKernel Configuration File System[0m.
19:49:11 [[0;32m  OK  [0m] Finished [0;1;39mColdplug All udev Devices[0m.
19:49:11 [[0;32m  OK  [0m] Reached target [0;1;39mSystem Initialization[0m.
19:49:11          Starting [0;1;39mdracut initqueue hook[0m...
19:49:11          Starting [0;1;39mShow Plymouth Boot Screen[0m...
19:49:11 [[0;32m  OK  [0m] Started [0;1;39mShow Plymouth Boot Screen[0m.

19:49:11 [[0;32m  OK  [0m] Started [0;1;39mForward Password R…s to Plymouth Directory Watch[0m.

19:49:11 [[0;32m  OK  [0m] Reached target [0;1;39mPaths[0m.

19:49:11 [[0;32m  OK  [0m] Reached target [0;1;39mBasic System[0m.

19:49:11 [[0;32m  OK  [0m] Found device [0;1;39mQEMU_HARDDISK systemd.[0m.

19:49:11 [[0;32m  OK  [0m] Finished [0;1;39mdracut initqueue hook[0m.

19:49:11 [[0;32m  OK  [0m] Reached target [0;1;39mInitrd Root Device[0m.

19:49:11 [[0;32m  OK  [0m] Reached target [0;1;39mRemote File Systems (Pre)[0m.

19:49:11 [[0;32m  OK  [0m] Reached target [0;1;39mRemote File Systems[0m.

19:49:11          Starting [0;1;39mFile System Check on /dev/sda1[0m...

19:49:11 [[0;32m  OK  [0m] Finished [0;1;39mFile System Check on /dev/sda1[0m.

19:49:11          Mounting [0;1;39m/sysroot[0m...

19:49:11 [[0;32m  OK  [0m] Mounted [0;1;39m/sysroot[0m.

19:49:11 [[0;32m  OK  [0m] Reached target [0;1;39mInitrd Root File System[0m.

19:49:11          Starting [0;1;39mReload Configuration from the Real Root[0m...

19:49:11 [[0;32m  OK  [0m] Finished [0;1;39mReload Configuration from the Real Root[0m.

19:49:11 [[0;32m  OK  [0m] Reached target [0;1;39mInitrd File Systems[0m.

19:49:11 [[0;32m  OK  [0m] Reached target [0;1;39mInitrd Default Target[0m.

19:49:11          Starting [0;1;39mdracut pre-pivot and cleanup hook[0m...

19:49:11 [[0;32m  OK  [0m] Finished [0;1;39mdracut pre-pivot and cleanup hook[0m.

19:49:11          Starting [0;1;39mCleaning Up and Shutting Down Daemons[0m...

19:49:11 [[0;32m  OK  [0m] Stopped target [0;1;39mTimers[0m.

19:49:11 [[0;32m  OK  [0m] Stopped [0;1;39mdracut pre-pivot and cleanup hook[0m.

19:49:11 [[0;32m  OK  [0m] Stopped target [0;1;39mInitrd Default Target[0m.

19:49:11 [[0;32m  OK  [0m] Stopped target [0;1;39mBasic System[0m.

19:49:11 [[0;32m  OK  [0m] Stopped target [0;1;39mInitrd Root Device[0m.

19:49:11 [[0;32m  OK  [0m] Stopped target [0;1;39mPaths[0m.

19:49:11 [[0;32m  OK  [0m] Stopped target [0;1;39mRemote File Systems[0m.

19:49:11 [[0;32m  OK  [0m] Stopped target [0;1;39mRemote File Systems (Pre)[0m.

19:49:11 [[0;32m  OK  [0m] Stopped target [0;1;39mSlices[0m.

19:49:11 [[0;32m  OK  [0m] Stopped target [0;1;39mSockets[0m.

19:49:11 [[0;32m  OK  [0m] Stopped target [0;1;39mSystem Initialization[0m.

19:49:11 [[0;32m  OK  [0m] Stopped target [0;1;39mSwap[0m.

19:49:11 [[0;32m  OK  [0m] Stopped [0;1;39mdracut initqueue hook[0m.

19:49:11          Starting [0;1;39mPlymouth switch root service[0m...

19:49:11 [[0;32m  OK  [0m] Stopped [0;1;39mApply Kernel Variables[0m.

19:49:11 [[0;32m  OK  [0m] Stopped [0;1;39mCreate Volatile Files and Directories[0m.

19:49:11 [[0;32m  OK  [0m] Stopped target [0;1;39mLocal File Systems[0m.

19:49:11 [[0;32m  OK  [0m] Stopped [0;1;39mColdplug All udev Devices[0m.

19:49:11          Stopping [0;1;39mRule-based Manage…for Device Events and Files[0m...

19:49:11 [[0;32m  OK  [0m] Finished [0;1;39mPlymouth switch root service[0m.

19:49:11 [[0;32m  OK  [0m] Stopped [0;1;39mRule-based Manager for Device Events and Files[0m.

19:49:11 [[0;32m  OK  [0m] Finished [0;1;39mCleaning Up and Shutting Down Daemons[0m.

19:49:11 [[0;32m  OK  [0m] Closed [0;1;39mudev Control Socket[0m.

19:49:11 [[0;32m  OK  [0m] Closed [0;1;39mudev Kernel Socket[0m.

19:49:11 [[0;32m  OK  [0m] Stopped [0;1;39mdracut pre-udev hook[0m.

19:49:11 [[0;32m  OK  [0m] Stopped [0;1;39mdracut cmdline hook[0m.

19:49:11          Starting [0;1;39mCleanup udev Database[0m...

19:49:11 [[0;32m  OK  [0m] Stopped [0;1;39mCreate Static Device Nodes in /dev[0m.

19:49:11 [[0;32m  OK  [0m] Stopped [0;1;39mCreate list of sta… nodes for the current kernel[0m.

19:49:11 [[0;32m  OK  [0m] Finished [0;1;39mCleanup udev Database[0m.

19:49:11 [[0;32m  OK  [0m] Reached target [0;1;39mSwitch Root[0m.

19:49:11          Starting [0;1;39mSwitch Root[0m...

19:49:11 

19:49:11 Welcome to [0;31mCentOS Linux 7 (Core)[0m!

19:49:11 

19:49:11 [[0;32m  OK  [0m] Stopped [0;1;39minitrd-switch-root.service[0m.

19:49:11 [[0;32m  OK  [0m] Created slice [0;1;39msystem-modprobe.slice[0m.

19:49:11 [[0;32m  OK  [0m] Created slice [0;1;39msystem-serial\x2dgetty.slice[0m.

19:49:11 [[0;32m  OK  [0m] Created slice [0;1;39muser.slice[0m.

19:49:11 [[0;32m  OK  [0m] Stopped [0;1;39msystemd-ask-password-plymouth.path[0m.

19:49:11 [[0;32m  OK  [0m] Started [0;1;39msystemd-ask-password-wall.path[0m.

19:49:11 [[0;32m  OK  [0m] Set up automount [0;1;39mproc-sys-fs-binfmt_misc.automount[0m.

19:49:11 [[0;32m  OK  [0m] Stopped target [0;1;39minitrd-switch-root.target[0m.

19:49:11 [[0;32m  OK  [0m] Stopped target [0;1;39minitrd-fs.target[0m.

19:49:11 [[0;32m  OK  [0m] Stopped target [0;1;39minitrd-root-fs.target[0m.

19:49:11 [[0;32m  OK  [0m] Reached target [0;1;39mslices.target[0m.

19:49:11 [[0;32m  OK  [0m] Reached target [0;1;39mswap.target[0m.

19:49:11 [[0;32m  OK  [0m] Listening on [0;1;39msystemd-coredump.socket[0m.

19:49:11 [[0;32m  OK  [0m] Listening on [0;1;39msystemd-initctl.socket[0m.

19:49:11 [[0;32m  OK  [0m] Listening on [0;1;39msystemd-udevd-control.socket[0m.

19:49:11 [[0;32m  OK  [0m] Listening on [0;1;39msystemd-udevd-kernel.socket[0m.

19:49:11          Mounting [0;1;39mdev-hugepages.mount[0m...

19:49:11          Mounting [0;1;39mdev-mqueue.mount[0m...

19:49:11          Mounting [0;1;39msys-kernel-debug.mount[0m...

19:49:11          Mounting [0;1;39mtmp.mount[0m...

19:49:11          Starting [0;1;39mkmod-static-nodes.service[0m...

19:49:11 [[0;32m  OK  [0m] Stopped [0;1;39msystemd-fsck-root.service[0m.
19:49:11 [[0;32m  OK  [0m] Stopped [0;1;39msystemd-journald.service[0m.
19:49:11          Starting [0;1;39msystemd-journald.service[0m...
19:49:11          Starting [0;1;39msystemd-remount-fs.service[0m...
19:49:11          Starting [0;1;39msystemd-sysctl.service[0m...
19:49:11          Starting [0;1;39msystemd-udev-trigger.service[0m...
19:49:11 [[0;32m  OK  [0m] Stopped [0;1;39msystemd-vconsole-setup.service[0m.
19:49:11 [[0;32m  OK  [0m] Stopped [0;1;39mplymouth-start.service[0m.
19:49:11 [[0;32m  OK  [0m] Mounted [0;1;39mdev-hugepages.mount[0m.
19:49:11 [[0;32m  OK  [0m] Mounted [0;1;39mdev-mqueue.mount[0m.
19:49:11 [[0;32m  OK  [0m] Mounted [0;1;39msys-kernel-debug.mount[0m.
19:49:11 [[0;32m  OK  [0m] Mounted [0;1;39mtmp.mount[0m.
19:49:11 [[0;32m  OK  [0m] Finished [0;1;39mkmod-static-nodes.service[0m.
19:49:11 [[0;32m  OK  [0m] Started [0;1;39msystemd-journald.service[0m.
19:49:11 [[0;32m  OK  [0m] Finished [0;1;39mmodprobe@drm.service[0m.
19:49:11 [[0;32m  OK  [0m] Finished [0;1;39msystemd-remount-fs.service[0m.
19:49:11 [[0;32m  OK  [0m] Finished [0;1;39msystemd-sysctl.service[0m.
19:49:11 [[0;32m  OK  [0m] Reached target [0;1;39mcryptsetup.target[0m.
19:49:11 [[0;32m  OK  [0m] Reached target [0;1;39mpaths.target[0m.
19:49:11          Starting [0;1;39msystemd-journal-flush.service[0m...
19:49:11          Starting [0;1;39msystemd-random-seed.service[0m...
19:49:11          Starting [0;1;39msystemd-sysusers.service[0m...
19:49:11 [[0;32m  OK  [0m] Finished [0;1;39msystemd-random-seed.service[0m.
19:49:11 [[0;32m  OK  [0m] Finished [0;1;39msystemd-udev-trigger.service[0m.
19:49:11 [[0;32m  OK  [0m] Finished [0;1;39msystemd-sysusers.service[0m.
19:49:11          Starting [0;1;39msystemd-tmpfiles-setup-dev.service[0m...
19:49:11 [[0;32m  OK  [0m] Finished [0;1;39msystemd-tmpfiles-setup-dev.service[0m.
19:49:11 [[0;32m  OK  [0m] Finished [0;1;39msystemd-journal-flush.service[0m.
19:49:11 [[0;32m  OK  [0m] Reached target [0;1;39mlocal-fs-pre.target[0m.
19:49:11 [[0;32m  OK  [0m] Reached target [0;1;39mlocal-fs.target[0m.
19:49:11          Starting [0;1;39mldconfig.service[0m...
19:49:11          Starting [0;1;39msystemd-tmpfiles-setup.service[0m...
19:49:11          Starting [0;1;39msystemd-udevd.service[0m...
19:49:11 [[0;32m  OK  [0m] Finished [0;1;39msystemd-tmpfiles-setup.service[0m.
19:49:11          Starting [0;1;39msystemd-update-utmp.service[0m...
19:49:11 [[0;32m  OK  [0m] Finished [0;1;39msystemd-update-utmp.service[0m.
19:49:11 [[0;32m  OK  [0m] Started [0;1;39msystemd-udevd.service[0m.
19:49:11 [[0;32m  OK  [0m] Finished [0;1;39mldconfig.service[0m.
19:49:11          Starting [0;1;39msystemd-update-done.service[0m...
19:49:11 [[0;32m  OK  [0m] Finished [0;1;39msystemd-update-done.service[0m.
19:49:11 [[0;32m  OK  [0m] Reached target [0;1;39msysinit.target[0m.
19:49:11 [[0;32m  OK  [0m] Started [0;1;39msystemd-tmpfiles-clean.timer[0m.
19:49:11 [[0;32m  OK  [0m] Reached target [0;1;39mtimers.target[0m.
19:49:11 [[0;32m  OK  [0m] Listening on [0;1;39mdbus.socket[0m.
19:49:11 [[0;32m  OK  [0m] Reached target [0;1;39msockets.target[0m.
19:49:11 [[0;32m  OK  [0m] Reached target [0;1;39mbasic.target[0m.
19:49:11          Starting [0;1;39mrc-local.service[0m...
19:49:11          Starting [0;1;39msystemd-logind.service[0m...
19:49:11          Starting [0;1;39msystemd-user-sessions.service[0m...
19:49:11          Starting [0;1;39mtestsuite-50.service[0m...
19:49:11 [[0;32m  OK  [0m] Started [0;1;39mrc-local.service[0m.
19:49:11 [[0;32m  OK  [0m] Finished [0;1;39msystemd-user-sessions.service[0m.
19:49:11 [[0;32m  OK  [0m] Started [0;1;39mserial-getty@ttyS0.service[0m.
19:49:11 [[0;32m  OK  [0m] Reached target [0;1;39mgetty.target[0m.
19:49:11 [[0;32m  OK  [0m] Started [0;1;39mdbus.service[0m.
19:49:11 [[0;32m  OK  [0m] Started [0;1;39msystemd-logind.service[0m.
19:49:11 [[0;32m  OK  [0m] Reached target [0;1;39mmulti-user.target[0m.
19:49:11          Starting [0;1;39msystemd-update-utmp-runlevel.service[0m...
19:49:11 [[0;32m  OK  [0m] Finished [0;1;39msystemd-update-utmp-runlevel.service[0m.
19:49:11 [[0;1;31mFAILED[0m] Failed to start [0;1;39mtestsuite-50.service[0m.
19:49:11 See 'systemctl status testsuite-50.service' for details.
19:49:11 [[0;32m  OK  [0m] Reached target [0;1;39mtestsuite.target[0m.
19:49:11          Starting [0;1;39mend.service[0m...
19:49:11 [[0;32m  OK  [0m] Stopped [0;1;39mend.service[0m.
19:49:11 [[0;32m  OK  [0m] Removed slice [0;1;39msystem-modprobe.slice[0m.
19:49:11 [[0;32m  OK  [0m] Stopped target [0;1;39mtestsuite.target[0m.
19:49:11 [[0;32m  OK  [0m] Stopped target [0;1;39mmulti-user.target[0m.
19:49:11 [[0;32m  OK  [0m] Stopped target [0;1;39mgetty.target[0m.
19:49:11 [[0;32m  OK  [0m] Stopped target [0;1;39mtimers.target[0m.
19:49:11 [[0;32m  OK  [0m] Stopped [0;1;39msystemd-tmpfiles-clean.timer[0m.
19:49:11          Stopping [0;1;39mdbus.service[0m...
19:49:11          Stopping [0;1;39mserial-getty@ttyS0.service[0m...
19:49:11          Stopping [0;1;39msystemd-logind.service[0m...
19:49:11          Stopping [0;1;39msystemd-random-seed.service[0m...
19:49:11 [[0;32m  OK  [0m] Stopped [0;1;39mserial-getty@ttyS0.service[0m.
19:49:11 [[0;32m  OK  [0m] Stopped [0;1;39msystemd-logind.service[0m.
19:49:11 [[0;32m  OK  [0m] Stopped [0;1;39mdbus.service[0m.
19:49:11 [[0;32m  OK  [0m] Stopped [0;1;39msystemd-random-seed.service[0m.
19:49:11 [[0;32m  OK  [0m] Removed slice [0;1;39msystem-serial\x2dgetty.slice[0m.
19:49:11 [[0;32m  OK  [0m] Stopped [0;1;39mrc-local.service[0m.
19:49:11          Stopping [0;1;39msystemd-user-sessions.service[0m...
19:49:11 [[0;32m  OK  [0m] Stopped [0;1;39msystemd-user-sessions.service[0m.
19:49:11 [[0;32m  OK  [0m] Stopped target [0;1;39mbasic.target[0m.
19:49:11 [[0;32m  OK  [0m] Stopped target [0;1;39mpaths.target[0m.
19:49:11 [[0;32m  OK  [0m] Stopped target [0;1;39mslices.target[0m.
19:49:11 [[0;32m  OK  [0m] Removed slice [0;1;39muser.slice[0m.
19:49:11 [[0;32m  OK  [0m] Stopped target [0;1;39msockets.target[0m.
19:49:11 [[0;32m  OK  [0m] Closed [0;1;39mdbus.socket[0m.
19:49:11 [[0;32m  OK  [0m] Stopped target [0;1;39msysinit.target[0m.
19:49:11 [[0;32m  OK  [0m] Stopped target [0;1;39mcryptsetup.target[0m.
19:49:11 [[0;32m  OK  [0m] Stopped [0;1;39msystemd-ask-password-wall.path[0m.
19:49:11 [[0;32m  OK  [0m] Stopped [0;1;39msystemd-sysctl.service[0m.
19:49:11 [[0;32m  OK  [0m] Stopped [0;1;39msystemd-update-done.service[0m.
19:49:11 [[0;32m  OK  [0m] Stopped [0;1;39mldconfig.service[0m.
19:49:11          Stopping [0;1;39msystemd-update-utmp.service[0m...
19:49:11 [[0;32m  OK  [0m] Stopped [0;1;39msystemd-update-utmp.service[0m.
19:49:11 [[0;32m  OK  [0m] Stopped [0;1;39msystemd-tmpfiles-setup.service[0m.
19:49:11 [[0;32m  OK  [0m] Stopped target [0;1;39mlocal-fs.target[0m.
19:49:11 [[0;32m  OK  [0m] Stopped target [0;1;39mlocal-fs-pre.target[0m.
19:49:11          Unmounting [0;1;39mtmp.mount[0m...
19:49:11 [[0;32m  OK  [0m] Stopped [0;1;39msystemd-tmpfiles-setup-dev.service[0m.
19:49:11 [[0;32m  OK  [0m] Stopped [0;1;39msystemd-sysusers.service[0m.
19:49:11 [[0;32m  OK  [0m] Stopped [0;1;39msystemd-remount-fs.service[0m.
19:49:11 [[0;32m  OK  [0m] Unmounted [0;1;39mtmp.mount[0m.
19:49:11 [[0;32m  OK  [0m] Stopped target [0;1;39mswap.target[0m.
19:49:11 [[0;32m  OK  [0m] Reached target [0;1;39mshutdown.target[0m.
19:49:11 [[0;32m  OK  [0m] Reached target [0;1;39mumount.target[0m.
19:49:11 [[0;32m  OK  [0m] Reached target [0;1;39mfinal.target[0m.
19:49:11 [[0;32m  OK  [0m] Finished [0;1;39msystemd-poweroff.service[0m.
19:49:11 [[0;32m  OK  [0m] Reached target [0;1;39mpoweroff.target[0m.
19:49:11 [    8.483246] Power down.
19:49:11 Spawning getter /root/systemd/build/journalctl -o export -D /var/tmp/systemd-test-TEST-50-DISSECT/root/var/log/journal/fd39d02e046a4015a9fce92fc8f54570...
19:49:11 Finishing after writing 3495 entries
19:49:11 Spawning getter /root/systemd/build/journalctl -o export -D /var/tmp/systemd-test-TEST-50-DISSECT/root/var/log/journal/remote...
19:49:11 Finishing after writing 0 entries
19:49:11 -rw-r----- 1 root root 8388608 Aug 31 18:49 /var/tmp/systemd-test-TEST-50-DISSECT/system.journal
19:49:11 QEMU test failed
19:49:11 make: *** [run] Error 1
19:49:11 make: Leaving directory `/root/systemd/test/TEST-50-DISSECT'
19:49:11 [TASK END] Mon 31 Aug 18:49:11 BST 2020
19:49:11 [RESULT] TEST-50-DISSECT - FAIL (EC: 2) (log file: /root/testsuite-logs-upstream.rXm/TEST-50-DISSECT_FAIL.log)
```
</details>

**Comment by bluca** [comment]
> > If and only if future measurements show a meaningful impact, would it be acceptable if I sent a follow-up to add a meson option to reinstate linking as a disabled-by-default alternative?
> 
> Can you first do the measurements? That'd be super interesting on its own. It is possible that this doesn't actually affect performance, and it'd be great to have a confirmation. But before you start working on a patch, I'd like to discuss how to we handle those issues in all the dlopen paths.

Certainly, it was a pure hypothetical point depending on what results would look like. It will probably be a couple of weeks before I can get around to it, but it's on my todo list.

**Comment by poettering** [comment]
> I'm a bit concerned about perf impact, given as far as I can see pid1 will not dlopen, but the forked processes that start each service will have to do it themselves, so the cost has to be incurred every time as far as I can see, and this is in my critical path. If and only if future measurements show a meaningful impact, would it be acceptable if I sent a follow-up to add a meson option to reinstate linking as a disabled-by-default alternative?

I doubt this would be measurable really. I mean, sure it will be slower, but the thing is: for the RootImage= stuff (and related stuff) we set up loopback block devices and wait for them to go through uevents/udev/rules/… and so on. That part is *really* slow, no doubt, and in cotrast to that looking up a few symbols should not matter. In particular as looking up the symbols needs to be done anyway, regardless if we link against this dynamically or via dlopen(). What might be a bit more expensive is the preparation for the first symbol lookup as we delay finding libcryptsetup.so into the forked child — but see it the other way for that: unless people actually use RootImage= (and I guess on most installations that's probably going to be the exception) things are quicker now, since we won't look for libcryptsetup anymore.

But of course, if this turns out to be a real problem, we can definitely revisit this, and optimize how things behave here.

**Comment by poettering** [comment]
> I'm also concerned about long-term maintainability. We copy the function headers and so-version suffix into our code. If things change, we won't notice. So longer-term I think we'll need to change how we do this. OTOH, short-term this shaves the deps in very important ways, so there's a clear immediate benefit for users.

Yeah, I wished we had a better story for this kind of stuff on Linux, i.e. "graceful" dynamic linking, where everything works like it always does except that deps listed in ELF headers can be marked as optional, in which case the binary is started but the symbols would all resolve to NULL, like a weak symbol if the dep is not found.

@fweimer is there any plan in glibc/ld.so/binutils/ELF to one day have something like that? As I understand MacOS supports dynamic linking like that, so it would have plenty prior art?

**Comment by keszybz** [comment]
centos-ci failed in the exact same way. Maybe we are not pulling in the right deps in the test image?

**Comment by keszybz** [comment]
```console
$ journalctl --file systemd-centos-ci/artifacts_a5bTmE/testsuite-logs-upstream.ypB/TEST-50-DISSECT/system.journal --no-hostname -o cat -t testsuite-50.sh
+ set -o pipefail
+ export SYSTEMD_LOG_LEVEL=debug
+ SYSTEMD_LOG_LEVEL=debug
+ cd /tmp
++ mktemp -d -t -p /tmp tmp.XXXXXX
+ image_dir=/tmp/tmp.Lk2yzG
+ '[' -z /tmp/tmp.Lk2yzG ']'
+ '[' '!' -d /tmp/tmp.Lk2yzG ']'
+ trap cleanup EXIT
+ cp /usr/share/minimal.raw /usr/share/minimal.roothash /usr/share/minimal.verity /tmp/tmp.Lk2yzG/
+ image=/tmp/tmp.Lk2yzG/minimal
++ cat /tmp/tmp.Lk2yzG/minimal.roothash
+ roothash=5c347ad0f70443b4e24f4a001eb7fb9a95ab6b9900ade5920164ad92b421c114
++ test -e /etc/os-release
++ echo /etc/os-release
+ os_release=/etc/os-release
+ systemd-dissect --json=short /tmp/tmp.Lk2yzG/minimal.raw
+ grep -q -F '{"rw":"ro","designator":"root","partition_uuid":null,"fstype":"squashfs","architecture":null,"verity":"external"'
Waiting for device (parent + 0 partitions) to appear...
sd-device-monitor: Unable to get network namespace of udev netlink socket, unable to determine if we are in host netns: Operation not supported
Successfully forked off '(sd-dissect)' as PID 509.
Mounting squashfs on /tmp/dissect-LSY7Ym (MS_RDONLY|MS_NODEV "")...
Failed to read /etc/hostname file of image, ignoring: No such file or directory
Failed to read /etc/machine-id file of image, ignoring: No such file or directory
Failed to read /etc/hostname: No such file or directory
Failed to read /etc/machine-info file of image, ignoring: No such file or directory
/etc/machine-id file is empty.
(sd-dissect) succeeded.
+ systemd-dissect /tmp/tmp.Lk2yzG/minimal.raw
+ grep -q -F MARKER=1
Waiting for device (parent + 0 partitions) to appear...
sd-device-monitor: Unable to get network namespace of udev netlink socket, unable to determine if we are in host netns: Operation not supported
Successfully forked off '(sd-dissect)' as PID 517.
Mounting squashfs on /tmp/dissect-Z5b9XQ (MS_RDONLY|MS_NODEV "")...
Failed to read /etc/hostname file of image, ignoring: No such file or directory
Failed to read /etc/machine-id file of image, ignoring: No such file or directory
Failed to read /etc/machine-info file of image, ignoring: No such file or directory
Failed to read /etc/hostname: No such file or directory
/etc/machine-id file is empty.
(sd-dissect) succeeded.
+ systemd-dissect /tmp/tmp.Lk2yzG/minimal.raw
+ grep -q -F -f /etc/os-release
Waiting for device (parent + 0 partitions) to appear...
Successfully forked off '(sd-dissect)' as PID 521.
Mounting squashfs on /tmp/dissect-hcEarS (MS_RDONLY|MS_NODEV "")...
Failed to read /etc/hostname file of image, ignoring: No such file or directory
Failed to read /etc/machine-id file of image, ignoring: No such file or directory
Failed to read /etc/machine-info file of image, ignoring: No such file or directory
Failed to read /etc/hostname: No such file or directory
/etc/machine-id file is empty.
(sd-dissect) succeeded.
+ mv /tmp/tmp.Lk2yzG/minimal.verity /tmp/tmp.Lk2yzG/minimal.fooverity
+ mv /tmp/tmp.Lk2yzG/minimal.roothash /tmp/tmp.Lk2yzG/minimal.foohash
+ systemd-dissect --json=short /tmp/tmp.Lk2yzG/minimal.raw --root-hash=5c347ad0f70443b4e24f4a001eb7fb9a95ab6b9900ade5920164ad92b421c114 --verity-data=/tmp/tmp.Lk2yzG/minimal.fo>
+ grep -q -F '{"rw":"ro","designator":"root","partition_uuid":null,"fstype":"squashfs","architecture":null,"verity":"external"'
Waiting for device (parent + 0 partitions) to appear...
sd-device-monitor: Unable to get network namespace of udev netlink socket, unable to determine if we are in host netns: Operation not supported
Successfully forked off '(sd-dissect)' as PID 527.
Mounting squashfs on /tmp/dissect-rATRAY (MS_RDONLY|MS_NODEV "")...
Failed to read /etc/hostname file of image, ignoring: No such file or directory
Failed to read /etc/machine-id file of image, ignoring: No such file or directory
Failed to read /etc/machine-info file of image, ignoring: No such file or directory
Failed to read /etc/hostname: No such file or directory
/etc/machine-id file is empty.
(sd-dissect) succeeded.
+ systemd-dissect /tmp/tmp.Lk2yzG/minimal.raw --root-hash=5c347ad0f70443b4e24f4a001eb7fb9a95ab6b9900ade5920164ad92b421c114 --verity-data=/tmp/tmp.Lk2yzG/minimal.fooverity
+ grep -q -F MARKER=1
Waiting for device (parent + 0 partitions) to appear...
Successfully forked off '(sd-dissect)' as PID 531.
Mounting squashfs on /tmp/dissect-j5BRe1 (MS_RDONLY|MS_NODEV "")...
Failed to read /etc/hostname file of image, ignoring: No such file or directory
Failed to read /etc/hostname: No such file or directory
Failed to read /etc/machine-id file of image, ignoring: No such file or directory
/etc/machine-id file is empty.
Failed to read /etc/machine-info file of image, ignoring: No such file or directory
(sd-dissect) succeeded.
+ systemd-dissect /tmp/tmp.Lk2yzG/minimal.raw --root-hash=5c347ad0f70443b4e24f4a001eb7fb9a95ab6b9900ade5920164ad92b421c114 --verity-data=/tmp/tmp.Lk2yzG/minimal.fooverity
+ grep -q -F -f /etc/os-release
Waiting for device (parent + 0 partitions) to appear...
sd-device-monitor: Unable to get network namespace of udev netlink socket, unable to determine if we are in host netns: Operation not supported
Successfully forked off '(sd-dissect)' as PID 535.
Mounting squashfs on /tmp/dissect-fyZwT2 (MS_RDONLY|MS_NODEV "")...
Failed to read /etc/hostname file of image, ignoring: No such file or directory
Failed to read /etc/machine-id file of image, ignoring: No such file or directory
Failed to read /etc/machine-info file of image, ignoring: No such file or directory
Failed to read /etc/hostname: No such file or directory
/etc/machine-id file is empty.
(sd-dissect) succeeded.
+ mv /tmp/tmp.Lk2yzG/minimal.fooverity /tmp/tmp.Lk2yzG/minimal.verity
+ mv /tmp/tmp.Lk2yzG/minimal.foohash /tmp/tmp.Lk2yzG/minimal.roothash
+ mkdir -p /tmp/tmp.Lk2yzG/mount /tmp/tmp.Lk2yzG/mount2
+ systemd-dissect --mount /tmp/tmp.Lk2yzG/minimal.raw /tmp/tmp.Lk2yzG/mount
Waiting for device (parent + 0 partitions) to appear...
Can't find symbol crypt_activate_by_signed_key: /lib64/libcryptsetup.so.12: undefined symbol: crypt_activate_by_signed_key
Failed to decrypt image: Accessing a corrupted shared library
+ cleanup
+ '[' -z /tmp/tmp.Lk2yzG ']'
+ rm -rf /tmp/tmp.Lk2yzG
```

File: src/shared/cryptsetup-util.c [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+        dl = dlopen("libcryptsetup.so.12", RTLD_LAZY);
+        if (!dl)
+                return log_debug_errno(SYNTHETIC_ERRNO(EOPNOTSUPP),
+                                       "libcryptsetup support is not installed: %s", dlerror());
+
+        r = dlsym_many_and_warn(
+                        dl,
+                        LOG_DEBUG,
+                        &sym_crypt_activate_by_passphrase, "crypt_activate_by_passphrase",
+                        &sym_crypt_activate_by_signed_key, "crypt_activate_by_signed_key",
```
Comment (bluca) [inline src/shared/cryptsetup-util.c:48]: I'm guessing this needs to be ifdef'd? ```#if HAVE_CRYPT_ACTIVATE_BY_SIGNED_KEY```

Comment (poettering) [reply src/shared/cryptsetup-util.c:48]: indeed

**Comment by poettering** [comment]
force pushed new version with the necessary ifdeffery fix. no other change. ptal.

**Review by bluca — APPROVED** [review]
LGTM

**Comment by poettering** [comment]
CI passed and both of you lgtm'ed it. So let's merge.

**Comment by fweimer** [comment]
> @fweimer is there any plan in glibc/ld.so/binutils/ELF to one day have something like that? As I understand MacOS supports dynamic linking like that, so it would have plenty prior art?

Our implementation of ELF dynamic linking supports weak symbols, but you have to link against a shim (or the oldest still-supported library version) to get run-time weak references. As a result, symbol versioning does not work correctly (and that's where we have a functionality gap).

`libcryptsetup` is actually a versioned library, but your `dlopen`-based wrapper does not cover that. It works fine as long as `libcryptsetup` does not provide multiple versions of the same symbol. Even today, weak symbols work under the same assumption.

Red Hat GmbH, https://de.redhat.com/ , Registered seat: Grasbrunn,
Commercial register: Amtsgericht Muenchen, HRB 153243,
Managing Directors: Charles Cachera, Brian Klemm, Laurie Krebs, Michael O'Neill

**Pull request merged**